File: exemplos/c/complexidade/exemplos_complexidade.c
/*
 * Exemplos de Algoritmos com Diferentes Complexidades
 * Professor: Vagner Cordeiro
 * Disciplina: ARA0174 - Algoritmos e Complexidade
 */

#include <stdio.h>
#include <stdlib.h>
#include <time.h>

// =================================
// ALGORITMOS O(1) - CONSTANTE
// =================================

// Acesso direto a elemento
int obter_primeiro(int arr[]) {
    return arr[0];  // Sempre 1 operação
}

// Operação aritmética simples
int soma_dois_numeros(int a, int b) {
    return a + b;   // Sempre 1 operação
}

// =================================
// ALGORITMOS O(n) - LINEAR
// =================================

// Busca linear
int busca_linear(int arr[], int n, int x) {
    for (int i = 0; i < n; i++) {
        if (arr[i] == x) {
            return i;
        }
    }
    return -1;
}

// Soma de array
int soma_array(int arr[], int n) {
    int soma = 0;
    for (int i = 0; i < n; i++) {
        soma += arr[i];
    }
    return soma;
}

// Impressão de array
void imprimir_array(int arr[], int n) {
    for (int i = 0; i < n; i++) {
        printf("%d ", arr[i]);
    }
    printf("\n");
}

// =================================
// ALGORITMOS O(log n) - LOGARÍTMICO
// =================================

// Busca binária (array deve estar ordenado)
int busca_binaria(int arr[], int esquerda, int direita, int x) {
    if (direita >= esquerda) {
        int meio = esquerda + (direita - esquerda) / 2;
        
        if (arr[meio] == x)
            return meio;
        
        if (arr[meio] > x)
            return busca_binaria(arr, esquerda, meio - 1, x);
        
        return busca_binaria(arr, meio + 1, direita, x);
    }
    return -1;
}

// Busca binária iterativa
int busca_binaria_iterativa(int arr[], int n, int x) {
    int esquerda = 0, direita = n - 1;
    
    while (esquerda <= direita) {
        int meio = esquerda + (direita - esquerda) / 2;
        
        if (arr[meio] == x)
            return meio;
        
        if (arr[meio] < x)
            esquerda = meio + 1;
        else
            direita = meio - 1;
    }
    return -1;
}

// =================================
// ALGORITMOS O(n²) - QUADRÁTICO
// =================================

// Bubble Sort
void bubble_sort(int arr[], int n) {
    for (int i = 0; i < n-1; i++) {
        for (int j = 0; j < n-i-1; j++) {
            if (arr[j] > arr[j+1]) {
                // Troca
                int temp = arr[j];
                arr[j] = arr[j+1];
                arr[j+1] = temp;
            }
        }
    }
}

// Selection Sort
void selection_sort(int arr[], int n) {
    for (int i = 0; i < n-1; i++) {
        int min_idx = i;
        for (int j = i+1; j < n; j++) {
            if (arr[j] < arr[min_idx]) {
                min_idx = j;
            }
        }
        // Troca
        int temp = arr[min_idx];
        arr[min_idx] = arr[i];
        arr[i] = temp;
    }
}

// Verificar se existem elementos duplicados
int tem_duplicatas(int arr[], int n) {
    for (int i = 0; i < n; i++) {
        for (int j = i + 1; j < n; j++) {
            if (arr[i] == arr[j]) {
                return 1;  // Tem duplicata
            }
        }
    }
    return 0;  // Não tem duplicata
}

// =================================
// ALGORITMOS O(n log n) - LINEARÍTMICO
// =================================

// Merge Sort - Função auxiliar para merge
void merge(int arr[], int esquerda, int meio, int direita) {
    int n1 = meio - esquerda + 1;
    int n2 = direita - meio;
    
    // Arrays temporários
    int* L = malloc(n1 * sizeof(int));
    int* R = malloc(n2 * sizeof(int));
    
    // Copiar dados para arrays temporários
    for (int i = 0; i < n1; i++)
        L[i] = arr[esquerda + i];
    for (int j = 0; j < n2; j++)
        R[j] = arr[meio + 1 + j];
    
    // Merge dos arrays temporários de volta em arr[]
    int i = 0, j = 0, k = esquerda;
    
    while (i < n1 && j < n2) {
        if (L[i] <= R[j]) {
            arr[k] = L[i];
            i++;
        } else {
            arr[k] = R[j];
            j++;
        }
        k++;
    }
    
    // Copiar elementos restantes
    while (i < n1) {
        arr[k] = L[i];
        i++;
        k++;
    }
    
    while (j < n2) {
        arr[k] = R[j];
        j++;
        k++;
    }
    
    free(L);
    free(R);
}

// Merge Sort principal
void merge_sort(int arr[], int esquerda, int direita) {
    if (esquerda < direita) {
        int meio = esquerda + (direita - esquerda) / 2;
        
        merge_sort(arr, esquerda, meio);
        merge_sort(arr, meio + 1, direita);
        
        merge(arr, esquerda, meio, direita);
    }
}

// =================================
// ALGORITMOS O(2^n) - EXPONENCIAL
// =================================

// Fibonacci recursivo (ineficiente)
int fibonacci_recursivo(int n) {
    if (n <= 1)
        return n;
    return fibonacci_recursivo(n-1) + fibonacci_recursivo(n-2);
}

// Torre de Hanói
void torre_hanoi(int n, char origem, char destino, char auxiliar) {
    if (n == 1) {
        printf("Mover disco 1 de %c para %c\n", origem, destino);
        return;
    }
    torre_hanoi(n-1, origem, auxiliar, destino);
    printf("Mover disco %d de %c para %c\n", n, origem, destino);
    torre_hanoi(n-1, auxiliar, destino, origem);
}

// =================================
// FUNÇÕES DE TESTE E MEDIÇÃO
// =================================

// Função para gerar array aleatório
void gerar_array_aleatorio(int arr[], int n, int max_valor) {
    for (int i = 0; i < n; i++) {
        arr[i] = rand() % max_valor;
    }
}

// Função para gerar array ordenado
void gerar_array_ordenado(int arr[], int n) {
    for (int i = 0; i < n; i++) {
        arr[i] = i + 1;
    }
}

// Função para copiar array
void copiar_array(int origem[], int destino[], int n) {
    for (int i = 0; i < n; i++) {
        destino[i] = origem[i];
    }
}

// Medir tempo de execução
double medir_tempo_ordenacao(void (*algoritmo)(int[], int), int arr[], int n) {
    clock_t inicio = clock();
    algoritmo(arr, n);
    clock_t fim = clock();
    return ((double)(fim - inicio)) / CLOCKS_PER_SEC;
}

// =================================
// FUNÇÃO PRINCIPAL E TESTES
// =================================

int main() {
    printf("=== EXEMPLOS DE COMPLEXIDADE DE ALGORITMOS ===\n\n");
    
    // Teste 1: Algoritmos O(1)
    printf("1. ALGORITMOS O(1) - CONSTANTE\n");
    int arr_teste[] = {5, 2, 8, 1, 9, 3};
    printf("Primeiro elemento: %d\n", obter_primeiro(arr_teste));
    printf("Soma 5 + 3: %d\n", soma_dois_numeros(5, 3));
    printf("\n");
    
    // Teste 2: Algoritmos O(n)
    printf("2. ALGORITMOS O(n) - LINEAR\n");
    int n = 6;
    printf("Array: ");
    imprimir_array(arr_teste, n);
    printf("Busca pelo 8: posição %d\n", busca_linear(arr_teste, n, 8));
    printf("Soma do array: %d\n", soma_array(arr_teste, n));
    printf("\n");
    
    // Teste 3: Algoritmos O(log n)
    printf("3. ALGORITMOS O(log n) - LOGARÍTMICO\n");
    int arr_ordenado[] = {1, 3, 5, 7, 9, 11, 13, 15};
    int n_ord = 8;
    printf("Array ordenado: ");
    imprimir_array(arr_ordenado, n_ord);
    printf("Busca binária pelo 7: posição %d\n", 
           busca_binaria(arr_ordenado, 0, n_ord-1, 7));
    printf("Busca binária pelo 6: posição %d\n", 
           busca_binaria(arr_ordenado, 0, n_ord-1, 6));
    printf("\n");
    
    // Teste 4: Algoritmos O(n²)
    printf("4. ALGORITMOS O(n²) - QUADRÁTICO\n");
    int arr_desordenado[] = {64, 34, 25, 12, 22, 11, 90};
    int n_des = 7;
    printf("Array original: ");
    imprimir_array(arr_desordenado, n_des);
    
    // Teste Bubble Sort
    int arr_bubble[7];
    copiar_array(arr_desordenado, arr_bubble, n_des);
    bubble_sort(arr_bubble, n_des);
    printf("Após Bubble Sort: ");
    imprimir_array(arr_bubble, n_des);
    
    // Teste Selection Sort
    int arr_selection[7];
    copiar_array(arr_desordenado, arr_selection, n_des);
    selection_sort(arr_selection, n_des);
    printf("Após Selection Sort: ");
    imprimir_array(arr_selection, n_des);
    printf("\n");
    
    // Teste 5: Algoritmos O(n log n)
    printf("5. ALGORITMOS O(n log n) - LINEARÍTMICO\n");
    int arr_merge[7];
    copiar_array(arr_desordenado, arr_merge, n_des);
    printf("Array original: ");
    imprimir_array(arr_merge, n_des);
    merge_sort(arr_merge, 0, n_des-1);
    printf("Após Merge Sort: ");
    imprimir_array(arr_merge, n_des);
    printf("\n");
    
    // Teste 6: Algoritmos O(2^n)
    printf("6. ALGORITMOS O(2^n) - EXPONENCIAL\n");
    printf("Fibonacci de 10: %d\n", fibonacci_recursivo(10));
    printf("Torre de Hanói para 3 discos:\n");
    torre_hanoi(3, 'A', 'C', 'B');
    printf("\n");
    
    // Teste de Performance
    printf("7. COMPARAÇÃO DE PERFORMANCE\n");
    int tamanhos[] = {100, 500, 1000};
    int num_testes = 3;
    
    for (int i = 0; i < num_testes; i++) {
        int size = tamanhos[i];
        int* arr = malloc(size * sizeof(int));
        int* arr_copia = malloc(size * sizeof(int));
        
        gerar_array_aleatorio(arr, size, 1000);
        
        // Teste Bubble Sort
        copiar_array(arr, arr_copia, size);
        double tempo_bubble = medir_tempo_ordenacao(bubble_sort, arr_copia, size);
        
        // Teste Selection Sort
        copiar_array(arr, arr_copia, size);
        double tempo_selection = medir_tempo_ordenacao(selection_sort, arr_copia, size);
        
        printf("Tamanho %d:\n", size);
        printf("  Bubble Sort: %.6f segundos\n", tempo_bubble);
        printf("  Selection Sort: %.6f segundos\n", tempo_selection);
        
        free(arr);
        free(arr_copia);
    }
    
    return 0;
}

File: exemplos/python/complexidade/exemplos_complexidade.py
"""
Exemplos de Algoritmos com Diferentes Complexidades
Professor: Vagner Cordeiro
Disciplina: ARA0174 - Algoritmos e Complexidade
"""

import time
import random
from typing import List, Tuple

# =================================
# ALGORITMOS O(1) - CONSTANTE
# =================================

def obter_primeiro(lista: List[int]) -> int:
    """Acesso direto a elemento - O(1)"""
    return lista[0]

def soma_dois_numeros(a: int, b: int) -> int:
    """Operação aritmética simples - O(1)"""
    return a + b

# =================================
# ALGORITMOS O(n) - LINEAR
# =================================

def busca_linear(lista: List[int], x: int) -> int:
    """Busca linear - O(n)"""
    for i in range(len(lista)):
        if lista[i] == x:
            return i
    return -1

def soma_lista(lista: List[int]) -> int:
    """Soma de lista - O(n)"""
    total = 0
    for elemento in lista:
        total += elemento
    return total

def imprimir_lista(lista: List[int]) -> None:
    """Impressão de lista - O(n)"""
    for elemento in lista:
        print(elemento, end=" ")
    print()

# =================================
# ALGORITMOS O(log n) - LOGARÍTMICO
# =================================

def busca_binaria_recursiva(lista: List[int], esquerda: int, direita: int, x: int) -> int:
    """Busca binária recursiva - O(log n)"""
    if direita >= esquerda:
        meio = esquerda + (direita - esquerda) // 2
        
        if lista[meio] == x:
            return meio
        elif lista[meio] > x:
            return busca_binaria_recursiva(lista, esquerda, meio - 1, x)
        else:
            return busca_binaria_recursiva(lista, meio + 1, direita, x)
    return -1

def busca_binaria_iterativa(lista: List[int], x: int) -> int:
    """Busca binária iterativa - O(log n)"""
    esquerda, direita = 0, len(lista) - 1
    
    while esquerda <= direita:
        meio = esquerda + (direita - esquerda) // 2
        
        if lista[meio] == x:
            return meio
        elif lista[meio] < x:
            esquerda = meio + 1
        else:
            direita = meio - 1
    return -1

# =================================
# ALGORITMOS O(n²) - QUADRÁTICO
# =================================

def bubble_sort(lista: List[int]) -> None:
    """Bubble Sort - O(n²)"""
    n = len(lista)
    for i in range(n-1):
        for j in range(n-i-1):
            if lista[j] > lista[j+1]:
                lista[j], lista[j+1] = lista[j+1], lista[j]

def selection_sort(lista: List[int]) -> None:
    """Selection Sort - O(n²)"""
    n = len(lista)
    for i in range(n-1):
        min_idx = i
        for j in range(i+1, n):
            if lista[j] < lista[min_idx]:
                min_idx = j
        lista[i], lista[min_idx] = lista[min_idx], lista[i]

def insertion_sort(lista: List[int]) -> None:
    """Insertion Sort - O(n²)"""
    for i in range(1, len(lista)):
        chave = lista[i]
        j = i - 1
        while j >= 0 and lista[j] > chave:
            lista[j + 1] = lista[j]
            j -= 1
        lista[j + 1] = chave

def tem_duplicatas(lista: List[int]) -> bool:
    """Verifica duplicatas - O(n²)"""
    n = len(lista)
    for i in range(n):
        for j in range(i + 1, n):
            if lista[i] == lista[j]:
                return True
    return False

# =================================
# ALGORITMOS O(n log n) - LINEARÍTMICO
# =================================

def merge_sort(lista: List[int]) -> List[int]:
    """Merge Sort - O(n log n)"""
    if len(lista) <= 1:
        return lista
    
    meio = len(lista) // 2
    esquerda = merge_sort(lista[:meio])
    direita = merge_sort(lista[meio:])
    
    return merge(esquerda, direita)

def merge(esquerda: List[int], direita: List[int]) -> List[int]:
    """Função auxiliar para merge - O(n)"""
    resultado = []
    i = j = 0
    
    while i < len(esquerda) and j < len(direita):
        if esquerda[i] <= direita[j]:
            resultado.append(esquerda[i])
            i += 1
        else:
            resultado.append(direita[j])
            j += 1
    
    resultado.extend(esquerda[i:])
    resultado.extend(direita[j:])
    return resultado

def quick_sort(lista: List[int]) -> List[int]:
    """Quick Sort - O(n log n) médio, O(n²) pior caso"""
    if len(lista) <= 1:
        return lista
    
    pivot = lista[len(lista) // 2]
    esquerda = [x for x in lista if x < pivot]
    meio = [x for x in lista if x == pivot]
    direita = [x for x in lista if x > pivot]
    
    return quick_sort(esquerda) + meio + quick_sort(direita)

# =================================
# ALGORITMOS O(2^n) - EXPONENCIAL
# =================================

def fibonacci_recursivo(n: int) -> int:
    """Fibonacci recursivo ineficiente - O(2^n)"""
    if n <= 1:
        return n
    return fibonacci_recursivo(n-1) + fibonacci_recursivo(n-2)

def fibonacci_memo(n: int, memo: dict = None) -> int:
    """Fibonacci com memoização - O(n)"""
    if memo is None:
        memo = {}
    
    if n in memo:
        return memo[n]
    
    if n <= 1:
        return n
    
    memo[n] = fibonacci_memo(n-1, memo) + fibonacci_memo(n-2, memo)
    return memo[n]

def torre_hanoi(n: int, origem: str, destino: str, auxiliar: str) -> None:
    """Torre de Hanói - O(2^n)"""
    if n == 1:
        print(f"Mover disco 1 de {origem} para {destino}")
        return
    
    torre_hanoi(n-1, origem, auxiliar, destino)
    print(f"Mover disco {n} de {origem} para {destino}")
    torre_hanoi(n-1, auxiliar, destino, origem)

# =================================
# FUNÇÕES DE UTILIDADE
# =================================

def gerar_lista_aleatoria(n: int, max_valor: int = 1000) -> List[int]:
    """Gera lista com números aleatórios"""
    return [random.randint(0, max_valor) for _ in range(n)]

def gerar_lista_ordenada(n: int) -> List[int]:
    """Gera lista ordenada"""
    return list(range(1, n + 1))

def medir_tempo(func, *args) -> Tuple[float, any]:
    """Mede tempo de execução de uma função"""
    inicio = time.time()
    resultado = func(*args)
    fim = time.time()
    return fim - inicio, resultado

def medir_tempo_ordenacao(algoritmo, lista: List[int]) -> float:
    """Mede tempo de algoritmo de ordenação (modifica a lista)"""
    inicio = time.time()
    algoritmo(lista)
    fim = time.time()
    return fim - inicio

# =================================
# ANÁLISE COMPARATIVA
# =================================

def comparar_algoritmos_busca(lista: List[int], elemento: int) -> None:
    """Compara algoritmos de busca"""
    print(f"Buscando elemento {elemento} em lista de tamanho {len(lista)}")
    
    # Busca linear
    tempo, pos = medir_tempo(busca_linear, lista, elemento)
    print(f"Busca Linear: {tempo:.6f}s, posição: {pos}")
    
    # Busca binária (lista deve estar ordenada)
    lista_ordenada = sorted(lista)
    tempo, pos = medir_tempo(busca_binaria_iterativa, lista_ordenada, elemento)
    print(f"Busca Binária: {tempo:.6f}s, posição: {pos}")

def comparar_algoritmos_ordenacao(tamanhos: List[int]) -> None:
    """Compara algoritmos de ordenação"""
    print("=== COMPARAÇÃO DE ALGORITMOS DE ORDENAÇÃO ===")
    
    for n in tamanhos:
        print(f"\nTamanho: {n}")
        lista_original = gerar_lista_aleatoria(n)
        
        # Bubble Sort
        lista_copia = lista_original.copy()
        tempo = medir_tempo_ordenacao(bubble_sort, lista_copia)
        print(f"Bubble Sort:    {tempo:.6f}s")
        
        # Selection Sort
        lista_copia = lista_original.copy()
        tempo = medir_tempo_ordenacao(selection_sort, lista_copia)
        print(f"Selection Sort: {tempo:.6f}s")
        
        # Insertion Sort
        lista_copia = lista_original.copy()
        tempo = medir_tempo_ordenacao(insertion_sort, lista_copia)
        print(f"Insertion Sort: {tempo:.6f}s")
        
        # Merge Sort
        tempo, _ = medir_tempo(merge_sort, lista_original)
        print(f"Merge Sort:     {tempo:.6f}s")
        
        # Quick Sort
        tempo, _ = medir_tempo(quick_sort, lista_original)
        print(f"Quick Sort:     {tempo:.6f}s")
        
        # Python sorted() (Timsort)
        tempo, _ = medir_tempo(sorted, lista_original)
        print(f"Python sorted(): {tempo:.6f}s")

def demonstrar_crescimento_fibonacci(max_n: int) -> None:
    """Demonstra diferença entre Fibonacci recursivo e com memoização"""
    print("=== FIBONACCI: RECURSIVO vs MEMOIZAÇÃO ===")
    
    for n in range(max_n + 1):
        # Fibonacci com memoização
        tempo_memo, resultado = medir_tempo(fibonacci_memo, n)
        
        # Fibonacci recursivo (apenas para n pequenos)
        if n <= 35:  # Limite para não demorar muito
            tempo_rec, _ = medir_tempo(fibonacci_recursivo, n)
            print(f"F({n}): {resultado}, Recursivo: {tempo_rec:.6f}s, Memo: {tempo_memo:.6f}s")
        else:
            print(f"F({n}): {resultado}, Memo: {tempo_memo:.6f}s (recursivo muito lento)")

# =================================
# FUNÇÃO PRINCIPAL
# =================================

def main():
    print("=== EXEMPLOS DE COMPLEXIDADE DE ALGORITMOS ===\n")
    
    # Demonstração 1: Algoritmos O(1)
    print("1. ALGORITMOS O(1) - CONSTANTE")
    lista_teste = [5, 2, 8, 1, 9, 3]
    print(f"Lista: {lista_teste}")
    print(f"Primeiro elemento: {obter_primeiro(lista_teste)}")
    print(f"Soma 5 + 3: {soma_dois_numeros(5, 3)}\n")
    
    # Demonstração 2: Algoritmos O(n)
    print("2. ALGORITMOS O(n) - LINEAR")
    print(f"Busca pelo 8: posição {busca_linear(lista_teste, 8)}")
    print(f"Soma da lista: {soma_lista(lista_teste)}\n")
    
    # Demonstração 3: Algoritmos O(log n)
    print("3. ALGORITMOS O(log n) - LOGARÍTMICO")
    lista_ordenada = [1, 3, 5, 7, 9, 11, 13, 15]
    print(f"Lista ordenada: {lista_ordenada}")
    print(f"Busca binária pelo 7: posição {busca_binaria_iterativa(lista_ordenada, 7)}")
    print(f"Busca binária pelo 6: posição {busca_binaria_iterativa(lista_ordenada, 6)}\n")
    
    # Demonstração 4: Algoritmos O(n²)
    print("4. ALGORITMOS O(n²) - QUADRÁTICO")
    lista_desordenada = [64, 34, 25, 12, 22, 11, 90]
    print(f"Lista original: {lista_desordenada}")
    
    lista_bubble = lista_desordenada.copy()
    bubble_sort(lista_bubble)
    print(f"Após Bubble Sort: {lista_bubble}")
    
    lista_selection = lista_desordenada.copy()
    selection_sort(lista_selection)
    print(f"Após Selection Sort: {lista_selection}\n")
    
    # Demonstração 5: Algoritmos O(n log n)
    print("5. ALGORITMOS O(n log n) - LINEARÍTMICO")
    resultado_merge = merge_sort(lista_desordenada)
    print(f"Merge Sort: {resultado_merge}")
    
    resultado_quick = quick_sort(lista_desordenada)
    print(f"Quick Sort: {resultado_quick}\n")
    
    # Demonstração 6: Algoritmos O(2^n)
    print("6. ALGORITMOS O(2^n) - EXPONENCIAL")
    print(f"Fibonacci de 10 (recursivo): {fibonacci_recursivo(10)}")
    print(f"Fibonacci de 10 (memoização): {fibonacci_memo(10)}")
    print("Torre de Hanói para 3 discos:")
    torre_hanoi(3, 'A', 'C', 'B')
    print()
    
    # Comparações de performance
    print("7. COMPARAÇÕES DE PERFORMANCE")
    
    # Comparar busca
    print("\n--- Comparação de Busca ---")
    lista_grande = gerar_lista_aleatoria(10000)
    elemento_busca = lista_grande[5000]  # Elemento no meio
    comparar_algoritmos_busca(lista_grande, elemento_busca)
    
    # Comparar ordenação
    comparar_algoritmos_ordenacao([100, 500, 1000])
    
    # Demonstrar crescimento Fibonacci
    print()
    demonstrar_crescimento_fibonacci(10)

if __name__ == "__main__":
    main()
